asio/include/xtt/asio/server_context.inl

```diff
@@ -0,0 +1,403 @@
+                                                                                              handler);
+                                                            }));
+    }
+
+    template <typename Handler>
+    bool server_context::set_cert(Handler handler)
+    {
+        if (cert_ != cert_map_.end())
+            return true;
+
+        auto suite_spec = handshake_ctx_.get_suite_spec();
+        if (!suite_spec) {
+            async_send_error_msg([this, handler]()
+                                 {
+                                     this->ec_ = boost::system::error_code(static_cast<int>(return_code::UNKNOWN_SUITE_SPEC),
+                                                                                            get_xtt_category());
+                                     handler(this->ec_);
+                                 });
+            return false;
+        }
+
+        auto cert_it = cert_map_.find(*suite_spec);
+        if (cert_map_.end() != cert_it) {
+            cert_ = cert_it;
+
+            return true;
+        } else {
+            async_send_error_msg([this, handler]()
+                                 {
+                                     this->ec_ = boost::system::error_code(static_cast<int>(return_code::BAD_CERTIFICATE),
+                                                                           get_xtt_category());
+                                     handler(this->ec_);
+                                 });
+            return false;
+        }
+    }
+
+    template <typename GPKLookupCallback,
+              typename AssignIdCallback,
+              typename Handler>
+    void
+    server_context::async_buildserverattest(GPKLookupCallback async_lookup_gpk_callback,
+                                            AssignIdCallback async_assign_id_callback,
+                                            Handler handler)
+    {
+        if (!set_cert(handler)) {
+            return; // set_cert takes care of raising the callback
+        }
+
+        server_handshake_context::io_buffer io_buf;
+
+        return_code new_rc = handshake_ctx_.build_serverattest(io_buf,
+                                                               *cert_->second,
+                                                               cookie_ctx_);
+
+        async_run_state_machine(new_rc,
+                                io_buf,
+                                async_lookup_gpk_callback,
+                                async_assign_id_callback,
+                                handler);
+    }
+
+    template <typename GPKLookupCallback,
+              typename AssignIdCallback,
+              typename Handler>
+    void
+    server_context::async_preparseidclientattest(GPKLookupCallback async_lookup_gpk_callback,
+                                                 AssignIdCallback async_assign_id_callback,
+                                                 Handler handler)
+    {
+        if (!set_cert(handler)) {
```

Comment:
The lookups after the one in `async_buildserverattest` *should* all be extraneous (and that's the reason for the short-circuit at the top of `set_cert`). 

I kept calling `set_cert` just as a way of checking that the certificate has in fact been set, to make sure we don't dereference an iterator that's actually `cert_map.end()`. The `set_cert` calls in the other two places could just be `cert_ != cert_map_.end()`, which is exactly what they actually are (again, because of the short-circuit).

One reason I wanted to do it this way is if, in the future, we set this up to allow the cookie-based server handshake approach. There, the server context state machine may get constructed and used in the middle of the handshake, so the cert would have to be looked-up anew. But, there are many things that have to be adapted to allow that usage, so this is just premature.

Reply:
> I kept calling set_cert just as a way of checking that the certificate has in fact been set, to
> make sure we don't dereference an iterator that's actually cert_map.end().

That's what I was guessing.   I personally would take out them out (they shouldn't be needed; if there's a logic bug, let's crash and be forced to track down the mistake), but am ok with leaving them.